### ISSUE TITLE
Make ModuleList and ModuleDict generic

### ISSUE THREAD

**Issue by Yura52** [pytorch/pytorch#80821]
### 🚀 The feature, motivation and pitch

Currently, the containers `nn.ModuleList` and `nn.ModuleDict` are typing-unaware, i.e. given this:

```python
class A(nn.Module):
    def __init__(self):
        self.my_modules = nn.ModuleList([nn.Linear(1, 1) for _ in range(10)])
```

`self.my_modules[i]` is treated as `nn.Module`, not as `nn.Linear`. For example, VSCode complains about snippets like `function_that_expects_tensor(self.my_modules[i].weight`), because it thinks that `.weight` can be both `Tensor` and `nn.Module`.

What I propose:

```python
from collections.abc import MutableSequence
from typing import TypeVar

ModuleListValue = TypeVar('ModuleListValue', bound=nn.Module)

class ModuleList(Module, MutableSequence[ModuleListValue]):
    # now, some methods can be typed, e.g.:    
    def __getitem__(...) -> ModuleListValue:
        ...
    ...
```

For `nn.DictModule`, it is a bit more complicated, since there are two different patterns:
- (A) `dict`-like: the set of keys is not fixed, all values are modules of the same type (e.g. `nn.Linear`)
- (B) `TypedDict`-like: the set of keys is fixed, the values can be of different types (e.g. `{'linear': nn.Linear(...), 'relu': nn.ReLU}`).

(A) can be implemented similarly to the previous example:

```python
...
class ModuleDict(Module, MutableMapping[str, ModuleDictValue]):
    ...
```

In fact, this can cover (B) as well in a very limited way (by setting `ModuleDictValue=nn.Module`). And it is unclear to me how to implement the fully functioning (B), it looks like we need something like `TypedMutableMapping`, but there is no such thing in typing, so I would start with `MutableMapping` and add TypedModuleDict when it becomes technically possible.

**Comment by fx-carton** [comment]
In the C++ API, this would be nice to have: if you know all the modules in the list will be of the same type, a typed `ModuleList` would allow for the convenience of the `ModuleList` registering modules, while avoiding the trouble of implementing `.forward()` with the generic `ModuleList` (either trying all known possibilities with `.as<Type>()` or using `AnyModule`).

Of course this is only relevant to the C++ API which can compile-time check the types of the modules in the list, so maybe I should open a new issue.

**Comment by SsnL** [comment]
I'd argue that generic ModuleList and ModuleDict are fairly useful. At least for me personally, I often have a parent module containing multiple submodules of the same kind. It is very annoying that typing them as `ModuleDict` loses the type information. The ugly workaround I do is  
```py
class Parent(nn.Module):
    childes: Mapping[str, SubModule]  # but really is a ModuleDict
```

**Comment by function2-llx** [comment]
@ssnl Maybe as a better workaround, you can use `Mapping[str, SubModule] | ModuleDict`.

**Comment by bdvllrs** [comment]
An alternative could be to have some kind of `TypedKey` class that externally keeps track of the types and use this as `ModuleDict` keys. 

Here is a very schematic implementation:
```python
from collections.abc import Mapping                                                                                                                                                                             from typing import Generic, Type, TypeVar, overload, reveal_type                                                                                                                                                


Tco = TypeVar("Tco", bound=Module, covariant=True)
T = TypeVar("T")


class Module:
    pass


class TypedKey(Generic[T, Tco]):
    def __init__(self, type_: Type[Tco], name: T):
        self.type_ = type_
        self.name = name


class ModuleDict:
    def __init__(self, data: Mapping[str | TypedKey[str, Module], Module]):
        self.data = data

    @overload
    def __getitem__(self, key: str) -> Module: ...

    @overload
    def __getitem__(self, key: TypedKey[str, Tco]) -> Tco: ...

    def __getitem__(self, key: str | TypedKey[str, Tco]) -> Module | Tco:
        return self.data[key]

class Mod1(Module):
    pass


class Mod2(Module):
    pass


if __name__ == "__main__":
    d = ModuleDict({"mod1": Mod1(), "mod2": Mod2()})
    reveal_type(d["mod1"])  # Revealed type is Module

    kmod1 = TypedKey(Mod1, "mod1")
    kmod2 = TypedKey(Mod2, "mod2")
    d2 = ModuleDict({kmod1: Mod1(), kmod2: Mod2()})
    reveal_type(d2[kmod1])  # Revealed type is Mod1
    reveal_type(d2[kmod2])  # Revealed type is Mod2
```

The advantage is that `ModuleDict` can still contain different Module types and have correct type inference when we use a TypedKey.

**Comment by RickoNoNo3** [comment]
That is exactly what I need. Please consider to integrate it.

**Comment by man2machine** [comment]
Would be very useful if ModuleDict could be typed like TypedDict or dict[..., ...] and ModuleList like list[...]

**Comment by dest1n1s** [comment]
Hope this can be supported soon. Adding some inheritance with type generics will be very hopeful in creating well-typed PyTorch projects.

**Comment by randolf-scholz** [comment]
> ModuleListValue would only be useful for homogenous module lists, but it's not clear to me how often you are actually going to have a homogenous list. Maybe we need ModuleTuple lol

Generic `ModuleList` would be useful when all the members are instances of some abstract base class or protocol, for example in Normalizing Flows I may want to stack multiple invertible models 

```python
class BijectionList(nn.ModuleList):  # should be Sequence[Bijection]
     def forward(self, x: Tensor, /) -> Tensor:
        for transform in self:
            x = transform.forward(x)
        return x

    def inverse(self, y: Tensor, /) -> Tensor:
        for transform in reversed(self):
            y = transform.inverse(y)   # <- without generic type, type-check will raise missing attr
        return y
```

One should note that generic types are incompatible with torchscript, so likely it would be better to create a new class altogether, ideally one that subclasses `collections.abc.Sequence` and one that subclasses `collections.abc.Mapping`.

**Comment by Skylion007** [comment]
Would definitely be interested tn this, a lot of typing hints were actually hidden by JIT decorators; this can be fixed by #153726 that will expose these typing hits to code. Is it still incompatible with torch script?

**Comment by AndBoyS** [comment]
If you want typed ModuleList right now you can use this:

```
from collections.abc import Iterator
from typing import Generic, TypeVar, overload

from torch import nn

T = TypeVar("T", bound=nn.Module)


class TypedModuleList(Generic[T], nn.ModuleList):
    def __iter__(self) -> Iterator[T]:
        return super().__iter__()  # type: ignore[no-any-return]

    def append(self, module: T) -> "TypedModuleList[T]":  # type: ignore[override]
        return super().append(module)  # type: ignore[return-value]

    @overload
    def __getitem__(self, idx: slice) -> "TypedModuleList[T]": ...

    @overload
    def __getitem__(self, idx: int) -> T: ...

    def __getitem__(self, idx):  # type: ignore[no-untyped-def]
        return super().__getitem__(idx)

    def __setitem__(self, idx: int, module: T) -> None:  # type: ignore[override]
        super().__setitem__(idx, module)

```

**Comment by chengjilai** [comment]
Also hitting this with `ty` (astral.sh type checker) on PyTorch 2.11 / Python 3.14. When accessing `self.blocks[i].reset_parameters()` on a homogeneous `ModuleList[MyBlock]`, `ty` reports `call-non-callable` because `__getitem__` is typed as returning `Module` rather than the specific subclass. The `@overload` annotation in `container.py` explicitly says `__getitem__(self, idx: int) -> Module:`, which loses the actual type.
